### PR TITLE
Add ethstats service

### PR DESCRIPTION
**Pull request by atoulme**
##  PR description
This adds native support for ethstats to besu.
## Fixed Issue(s)

This fixes [BESU-191](https://jira.hyperledger.org/browse/BESU-191).

### CODE REVIEW

**Comment by atoulme** [comment]
This is missing the hooks to report new head, transactions count and history requests. I will provide those now. Any feedback on the construction is welcome.

**Comment by ajsutton** [comment]
Haven't looked at the details, but you've reverted the reference tests submodule again. :)

**Comment by atoulme** [comment]
The equivalent of having salad stuck in your teeth... OK that's fixed. Please take a look at leisure.

Please note this depends on a snapshot version of Apache Tuweni, so a release of Apache Tuweni is required.

**Comment by atoulme** [comment]
Sorry, the build won't pass right now - bad snapshot caching. May I get a review on how this is integrated with Besu? I didn't find a way to get the current hashrate and gas price.

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/ethstats/EthStatsService.java [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+            block.getBody().getTransactions(),
+            block.getBody().getOmmers(),
+            difficulty));
+  }
+
+  public void start(final Vertx vertx) {
+    Log4j2LoggerProvider provider = new Log4j2LoggerProvider();
+    URI ethStatsURI =
+        URI.create(
+            "ws://" + ethStatsParameters.getHost() + ":" + ethStatsParameters.getPort() + "/api");
```
Comment (nfmelendez) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/ethstats/EthStatsService.java]: i am not in the context of this code but using websocket without ssl seems something that might or not be important. just a  small warn.

**Comment by shemnon** [comment]
I've got EthDenver (on the tails of a tight deadline release) this weekend so I don't think I'll be able to give a quality review until Tuesday.

File: ethereum/eth/build.gradle [hunk 2]
```diff
@@ -43,7 +43,10 @@ dependencies {
   implementation 'com.google.guava:guava'
   implementation 'io.vertx:vertx-core'
   implementation 'org.apache.tuweni:tuweni-bytes'
+  implementation 'org.apache.tuweni:tuweni-ethstats'
   implementation 'org.apache.tuweni:tuweni-units'
+  implementation 'org.logl:logl-logl'
```
Comment (shemnon) [inline ethereum/eth/build.gradle]: Currently our logging library is Log4j2 and only Log4j2.  Can this PR be done without the Logl dependency?

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/ethstats/EthStatsService.java [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    // TODO provide client identification string
+    this.reporter =
+        new EthStatsReporter(
```
Comment (shemnon) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/ethstats/EthStatsService.java]: I'm not terribly excited about using an external library for what should be core functionality.  Would this make more sense as a third party plugin via the 1.4 plugins APIs?  Is there any stats we need to expose that are not already available via the plugin?

File: gradle/versions.gradle [hunk 4]
```diff
@@ -71,6 +71,7 @@ dependencyManagement {
     dependency 'org.apache.tuweni:tuweni-toml:0.10.0'
     dependency 'org.apache.tuweni:tuweni-units:0.10.0'
     dependency 'org.apache.tuweni:tuweni-net:0.10.0'
+    dependency 'org.apache.tuweni:tuweni-ethstats:1.0.0-SNAPSHOT'
```
Comment (shemnon) [inline gradle/versions.gradle]: We cant merge this with a snapshot dependency.

File: ethereum/eth/build.gradle (hunk 2 above)
Comment (atoulme) [reply ethereum/eth/build.gradle]: Logl proxies to log4j2.

File: gradle/versions.gradle (hunk 4 above)
Comment (atoulme) [reply gradle/versions.gradle]: Yes, I am refraining to cut 1.0.0 quite yet until you review. Chicken and egg issue.

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/ethstats/EthStatsService.java (hunk 3 above)
Comment (atoulme) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/ethstats/EthStatsService.java]: No idea, but I'm pretty sure there are quite a few things that you'd need to report, such as the mining hashrate, which are not currently exposed.

Maybe this is a good stopping point: either you want this in or as a plugin. What should it be?

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/ethstats/EthStatsService.java (hunk 1 above)
Comment (atoulme) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/ethstats/EthStatsService.java]: Good point. I don't remember how this is done on geth's side. I'll look.

Comment (atoulme) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/ethstats/EthStatsService.java]: Guess what, they try wss first and default back to ws if wss fails.

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/ethstats/EthStatsService.java (hunk 3 above)
Comment (atoulme) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/ethstats/EthStatsService.java]: I have moved this discussion to rocket.chat to a wider audience, we can discuss more there.

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/ethstats/EthStatsService.java (hunk 1 above)
Comment (atoulme) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/ethstats/EthStatsService.java]: This is resolved now.

**Comment by atoulme** [comment]
This PR is closing as the team indicated they would rather have integration done through the plugin system. The plugin system might need a few additions to accurately report to the same level done here, that can be arranged a bit later.

**Comment by atoulme** [comment]
See https://chat.hyperledger.org/channel/besu-contributors?msg=2g5n8tT27uPiYbWo5 for the discussion.

**Pull request closed without merging**